addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -3242,6 +3299,49 @@ registry.BackgroundImage = SnippetOptionWidget.extend({
             this.$target.trigger('background_changed', [previewMode]);
         }
     },
+    /**
+     * Toggles background image on or off.
+     *
+     * @see this.selectClass for parameters
+     */
+    toggleBgImage(previewMode, widgetValue, params) {
+        if (this._computeWidgetState('background')) {
+            return this.background(false, '', {});
+        } else {
+            const [bgImgPicker] = this._requestUserValueWidgets('bg_image_opt');
+            if (bgImgPicker) {
+                this.trigger_up('snippet_edition_request', {exec: () => bgImgPicker.el.click()});
+            }
+        }
```

Comment:
I could try to improve the trigger system, as it is very bare bones and quite hacky right now, I'm just not sure if doing it in this task is a great idea. It would probably be pretty involved. I'm thinking that widgets should only set their own values and trigger themselves through common methods on the `UserValueWidget` base class, this way both of those methods are common to all userValueWidgets and they have a common interface that's public and accessible to other widgets, then we can have a `data-trigger` that would do the same thing as a user-interaction (toggle a checkbox, open a select, open a media-dialog, ...), and a `data-set-widget-value="widgetName: widgetValue, otherWidget: otherValue"` that would do the same thing as selecting a specific value (eg setting a checkbox to being checked, picking a specific option in a select, selecting a specific image in the media-dialog, ...)

This sounds like it should work and be more flexible, but it would require rewriting a few parts of the current userValueWidgets. From what I understand we're trying to get this task merged ASAP so that the designers can work on shapes and integrate them into thems and stuff, so maybe this should better be done separately at a later point? While the "click on the imagepicker el" trick does sound very hacky, it's pretty much what the system I'm describing would achieve except without dispatching an event, so if we intend to rework the trigger system soon I think this hack is acceptable.

Reply:
Ok, but at least consider my first two points then